### ISSUE TITLE
Fine tune stopping point, early stopping, learning rate, best model, etc.

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#246]
I just trained an MMLV model on deterministic captions for 300 epochs, though I've seen some issues similar to what I'm about to describe when training for 500 epochs as well. Looking at these results partially inspired issue #244 but there is a bigger issue here. This is the training loss vs validation loss plot:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/c8587044-c61d-4d1f-a3e5-d86c03e18043" />

This looks like overfitting. The validation loss dips down to 0.32 then ticks back up to 0.35 while the training loss keeps going down. This doesn't really happen with MM-Full or MM-Simple, and definitely not with Mario, though in the Mario paper we did differentiate the "best" model from the final model in terms of caption score, though I think the code in train_diffusion.py is now set up to use validation loss when training with caption_keys. 

So, first try training with early stopping on and see what happens. When does training stop and how do the results look? For the sake of comparison, I've usually preferred to train for a fixed number of epochs and then use the best model encountered over that period (even if it was not the final model), but I'm honestly not fully sure how the code handles that any more (easy to forget with time and complexity) so look into that. If training for 300 epochs actually finds a better model at epoch 200, then not only should we save the best model instead, but some output in the model directory should make it very clear/explicit that this is what happened, and report which epoch the model came from.

However, more enticing is the idea that we could get the validation loss lower than we are seeing. This is where tweaking the learning rate scheduler could have an impact. There are so many things we could try, and each time we try something, it takes time, so I'm not sure how much effort to dedicate to this, but it is worth looking into.

**Comment by schrum2** [comment]
No changed code yet, but more evidence that fixing this is important. Did an MMLV run with just one caption per scene #239 and led to obvious overfitting:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/476ee01a-17f0-4d1d-9b17-34f2b1148d13" />

Doing an early stopping run now to see how well the existing code works.

**Comment by schrum2** [comment]
I tried early_stopping on an MMLV dataset with gemma captions and it ended way too early. Specifically, it ended at 50 epochs, but the results don't look great:

<img width="615" height="335" alt="Image" src="https://github.com/user-attachments/assets/521feb00-edfe-4cfe-9a9a-f69d5d72c6e1" />

<img width="611" height="330" alt="Image" src="https://github.com/user-attachments/assets/a8ab0b36-3ae7-4ae9-9045-99a024b0a0c8" />

<img width="611" height="344" alt="Image" src="https://github.com/user-attachments/assets/753f81ba-9697-4f3e-90b3-397a441872cb" />

This is especially bad because this was a training run where only one caption was associated with each scene, which should create a very clear connection between source scenes and generated scenes.

I could require a minimum number of epochs, but that seems like a bad idea. I would rather just fix the early stopping configuration.

**Comment by schrum2** [comment]
Making validation loss calculation deterministic could help. Also added plateau LR. Testing

**Comment by schrum2** [comment]
I'm still not liking the results with early stopping. I should try the plateau approach, though for the moment I'm just training for 200 epochs

**Comment by schrum2** [comment]
Some results from the lab seem to indicate that training with captions from multiple models (Qwen and Gemma) is less prone to overfitting than training with captions from just one. My training runs using only one caption per scene from a single model seem more prone to overfitting. They train ok for 200 epochs but validation loss goes up between 200 and 300.
